qlog/frame.go

```diff
@@ -0,0 +1,360 @@
+package qlog
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/lucas-clemente/quic-go/internal/protocol"
+	"github.com/lucas-clemente/quic-go/internal/wire"
+)
+
+type frame struct {
+	Frame interface{}
+}
+
+type streamType protocol.StreamType
+
+func (s streamType) String() string {
+	switch protocol.StreamType(s) {
+	case protocol.StreamTypeUni:
+		return "unidirectional"
+	case protocol.StreamTypeBidi:
+		return "bidirectional"
+	default:
+		panic("unknown stream type")
+	}
+}
+
+func escapeStr(str string) []byte { return []byte("\"" + str + "\"") }
+
+type connectionID protocol.ConnectionID
+
+func (c connectionID) MarshalJSON() ([]byte, error) {
+	return escapeStr(fmt.Sprintf("%x", c)), nil
+}
+
+type cryptoFrame struct {
+	Offset protocol.ByteCount
+	Length protocol.ByteCount
+}
+
+type streamFrame struct {
+	StreamID protocol.StreamID
+	Offset   protocol.ByteCount
+	Length   protocol.ByteCount
+	FinBit   bool
+}
+
+func transformFrame(wf wire.Frame) *frame {
+	// We don't want to store CRYPTO and STREAM frames, for multiple reasons:
```

Comment:
Why not make a copy of the struct and zero out the slices? That would allow you to reuse the structs?

Reply:
Because we still need the length of the frame. `wire.StreamFrame` does implement a `DataLen` method, but it just wraps `len(frame.Data)`.